### PR TITLE
Convert `ga-helper.js` to TypeScript

### PR DESCRIPTION
**Pull request by zekehuntergreen**
## What does this change?
- Converts `ga-helper.js` to TypeScript

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [ ] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [ ] Locally
- [x] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/video/ga-helper.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	play?: 'metric1';
+	skip?: 'metric2';
+	'25'?: 'metric3';
+	'50'?: 'metric4';
+	'75'?: 'metric5';
+	watched25?: 'metric3';
+	watched50?: 'metric4';
+	watched75?: 'metric5';
+	end?: 'metric6';
+};
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/video/ga-helper.ts:30]: If you want to keep this quite generic, you can use the [`Partial<T>`][] and [`Record<T>`][] utility types.


```suggestion
type EventMetrics = Partial<Record<EventType, Metrics>>
```

[`Partial<T>`]: https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype
[`Record<T>`]: https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type

File: static/src/javascripts/projects/common/modules/video/ga-helper.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+type InitialFieldsObject = {
+	eventCategory: string;
+	eventAction: string;
+	eventLabel: string;
+	dimension19: string;
+	dimension20: string;
+};
+type MetricFieldsObject = {
+	[key in Metrics]?: 1;
+};
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/video/ga-helper.ts:48]: Similar to above, if you want to use utility types.

```suggestion
type MetricFieldsObject = Record<Metrics, 1>
```

File: static/src/javascripts/projects/common/modules/video/ga-helper.ts (hunk 1 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/video/ga-helper.ts:30]: Alternatively, with a m
```suggestion
type EventMetrics = {
	[key in EventType]?: Metrics;
};
```

File: static/src/javascripts/projects/common/modules/video/ga-helper.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	| 'skip'
+	| '25'
+	| '50'
+	| '75'
+	| 'watched25'
+	| 'watched50'
+	| 'watched75'
+	| 'end';
+
+type Metrics =
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/video/ga-helper.ts:12]: I would tend to use the singular, here: `Metric`

**Review by mxdvl — APPROVED** [review]
Great one! Some comments on utility types, but no blockers.

Don’t forget to run `make commercial-graph` to build the latest deps for visualisation :)

File: static/src/javascripts/projects/common/modules/video/ga-helper.ts (hunk 1 above)
Comment (zekehuntergreen) [reply static/src/javascripts/projects/common/modules/video/ga-helper.ts:30]: Thanks for the suggestion! I don't think I knew you could do this.
From looking at the uses of `buildGoogleAnalyticsEvent` it seems like the `EventMetrics` s always conform to a pretty similar structure; `play` is always `metric1`, `skip` is `metric2`, etc. 
It seems to me that we might as well have our a types describe this structure as precisely as possible and disallow objects that don't match e.g. {'play': 'metric2'}. The downside is that we don't get to take advantage of `EventType` and `Metric` if we define `EventMetrics` from scratch 🙁

Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/video/ga-helper.ts:32]: Thinking about this a little more, you could do something like this ([TypeScript Playground][]):

```ts
const eventMetrics = {
	play: 'metric1',
	skip: 'metric2',
	'25': 'metric3',
	'50': 'metric4',
	'75': 'metric5',
	watched25: 'metric3',
	watched50: 'metric4',
	watched75: 'metric5',
	end: 'metric6',
} as const;

type EventKey = keyof typeof eventMetrics;
type Metric = typeof eventMetrics[EventKey];
```

[TypeScript Playground]: https://www.typescriptlang.org/play?#code/MYewdgzgLgBApgNzmKBZOUBOBLYEYC8MA3gFACQADgDYCGAngFwwDkAthjsAIwsA0FCAGtslZu064ATPwospAVhbiOWXAGZZ5FgoAMy1qq4AWLSwDsSlZOBKB5AO60owABZwAJoutrgm+04u7h56PiZagW6elmG4dhTIHrHAAGyyAL4wtPigkFAA3KSkUPSUcDAAokgoANJw9IQwQvUgAGYwJWVt8NVoNhCFneXovo1D3YjIfb4QANpVU3X0ALr5QA

Comment (zekehuntergreen) [reply static/src/javascripts/projects/common/modules/video/ga-helper.ts:32]: Yes, this is much better! Thanks for the suggestion
Made this change in #24386

**Pull request merged**